File: apps/runtime/src/git/repository/command.rs
//! The one place a Git child process is spawned for a repository command.

use super::*;

pub(super) fn sanitize(message: &str) -> String {
    let redacted = redact_secrets(message);
    let url = Regex::new(r"(?i)(https?|ssh)://[^/\s@]+@").expect("credential URL regex");
    url.replace_all(&redacted, "$1://[redacted]@")
        .chars()
        .take(8192)
        .collect()
}

pub(super) struct CommandOutput {
    pub(super) stdout: Vec<u8>,
    pub(super) stderr: Vec<u8>,
    pub(super) status: Option<i32>,
}
pub(super) fn command_error(output: &CommandOutput) -> AppError {
    let message = if output.stderr.is_empty() {
        &output.stdout
    } else {
        &output.stderr
    };
    AppError::Internal(format!(
        "Git operation failed: {}",
        sanitize(&String::from_utf8_lossy(message))
    ))
}

pub(super) async fn read_output(
    mut reader: impl AsyncRead + Unpin,
    limit: usize,
) -> AppResult<Vec<u8>> {
    let mut output = Vec::new();
    let mut chunk = [0u8; 8192];
    loop {
        let count = reader.read(&mut chunk).await?;
        if count == 0 {
            return Ok(output);
        }
        if count > limit - output.len() {
            return Err(AppError::Internal(
                "Git output exceeded its bounded budget".into(),
            ));
        }
        output.extend_from_slice(&chunk[..count]);
    }
}

/// Reads stderr while reporting the percentages it carries.
///
/// Git writes progress with carriage returns rather than newlines, so this
/// splits on both and looks at each completed line. The buffer is bounded the
/// same way the whole read is: a remote that never writes a separator must not
/// be able to grow it.
///
/// Without an observer this is [`read_output`] exactly, so the ordinary command
/// pays nothing for a feature only four of them use.
pub(super) async fn read_progress(
    mut reader: impl AsyncRead + Unpin,
    limit: usize,
    observer: Option<Arc<dyn Fn(u32) + Send + Sync>>,
) -> AppResult<Vec<u8>> {
    let Some(observer) = observer else {
        return read_output(reader, limit).await;
    };
    let mut output = Vec::new();
    let mut line = Vec::new();
    let mut chunk = [0u8; 8192];
    loop {
        let count = reader.read(&mut chunk).await?;
        if count == 0 {
            report_progress(&line, &observer);
            return Ok(output);
        }
        if count > limit - output.len() {
            return Err(AppError::Internal(
                "Git output exceeded its bounded budget".into(),
            ));
        }
        output.extend_from_slice(&chunk[..count]);
        for byte in &chunk[..count] {
            if matches!(*byte, b'\r' | b'\n') {
                report_progress(&line, &observer);
                line.clear();
            } else if line.len() < 4096 {
                line.push(*byte);
            }
        }
    }
}

/// `Receiving objects:  47% (470/1000)` -> 47.
///
/// Git prints a percentage for several phases in turn, each restarting at zero,
/// so this reports what the newest line said and nothing more. Deciding what to
/// do with a number that went down belongs to whoever renders it.
fn report_progress(line: &[u8], observer: &Arc<dyn Fn(u32) + Send + Sync>) {
    let Ok(text) = std::str::from_utf8(line) else {
        return;
    };
    let Some((before, _)) = text.split_once('%') else {
        return;
    };
    let digits: String = before.chars().rev().take_while(char::is_ascii_digit).collect();
    if digits.is_empty() {
        return;
    }
    if let Ok(percent) = digits.chars().rev().collect::<String>().parse::<u32>()
        && percent <= 100
    {
        observer(percent);
    }
}

pub(super) struct GitRunPolicy {
    pub(super) timeout: Duration,
    pub(super) allow_helpers: bool,
    /// Extra variables applied after the fixed environment below, so a caller
    /// can never weaken the prompt/askpass/config lockdown.
    pub(super) environment: Vec<(String, String)>,
    /// Called with each percentage `git --progress` writes to stderr, while it
    /// is still writing (Git 设计 §10, 进度).
    ///
    /// It exists for the network commands, which are the only ones that take
    /// long enough for a person to wonder whether anything is happening. It is
    /// a display value and nothing reads it back: the outcome still comes from
    /// the process's exit status, so a lost percentage costs a stalled bar and
    /// never a wrong result.
    pub(super) progress: Option<Arc<dyn Fn(u32) + Send + Sync>>,
}
pub(super) async fn run_git_status(
    directory: &Path,
    arguments: Vec<String>,
    policy: GitRunPolicy,
    token: &Cancellation,
    mutation_started: Option<&AtomicBool>,
    lease: &RepositoryCommandLease,
    caller: &Cancellation,
) -> AppResult<CommandOutput> {
    if token.requested.load(Ordering::SeqCst)
        || lease.inner.stopping.requested.load(Ordering::SeqCst)
        || caller.requested.load(Ordering::SeqCst)
    {
        return Err(AppError::Conflict("Git operation cancelled".into()));
    }
    let mut command = Command::new("git");
    command
        .args([
            "--no-pager",
            "-c",
            "core.quotepath=false",
            "-c",
            "color.ui=false",
            "-c",
            "i18n.logOutputEncoding=UTF-8",
        ])
        .args(arguments)
        .current_dir(directory)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .env("LC_ALL", "C")
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_ASKPASS", "")
        .env("SSH_ASKPASS", "")
        .env("GCM_INTERACTIVE", "never")
        .env("GIT_OPTIONAL_LOCKS", "0");
    for (name, _) in std::env::vars_os() {
        let key = name.to_string_lossy();
        if matches!(
            key.as_ref(),
            "GIT_DIR"
                | "GIT_WORK_TREE"
                | "GIT_COMMON_DIR"
                | "GIT_INDEX_FILE"
                | "GIT_OBJECT_DIRECTORY"
                | "GIT_ALTERNATE_OBJECT_DIRECTORIES"
                | "GIT_NAMESPACE"
                | "GIT_CONFIG"
                | "GIT_CONFIG_COUNT"
                | "GIT_CONFIG_PARAMETERS"
                | "GIT_EXTERNAL_DIFF"
                | "GIT_DIFF_OPTS"
                | "GIT_CURL_VERBOSE"
        ) || key.starts_with("GIT_CONFIG_KEY_")
            || key.starts_with("GIT_CONFIG_VALUE_")
            || key.starts_with("GIT_TRACE")
        {
            command.env_remove(name);
        }
    }
    if !policy.allow_helpers {
        crate::git::access::restrict_async(&mut command);
    }
    // Applied last so the lockdown above cannot be reintroduced by a caller,
    // and only ever the sequence editor an interactive rebase supplies.
    for (name, value) in &policy.environment {
        command.env(name, value);
    }
    let mut child = command
        .spawn()
        .map_err(|_| AppError::Internal("Could not start Git".into()))?;
    lease.started.store(true, Ordering::SeqCst);
    if let Some(started) = mutation_started {
        started.store(true, Ordering::SeqCst);
    }
    let stdout = child.stdout.take().ok_or_else(malformed)?;
    let stderr = child.stderr.take().ok_or_else(malformed)?;
    let observer = policy.progress.clone();
    let execution = async {
        let (stdout, stderr, status) = tokio::try_join!(
            read_output(stdout, MAX_OUTPUT),
            read_progress(stderr, MAX_STDERR, observer),
            async { child.wait().await.map_err(AppError::from) }
        )?;
        Ok(CommandOutput {
            stdout,
            stderr,
            status: status.code(),
        })
    };
    let result = tokio::select! {
        result = execution => result,
        _ = token.cancelled() => Err(AppError::Conflict("Git process was cancelled; verify repository/remote state before retrying".into())),
        _ = caller.cancelled() => Err(AppError::Conflict("Git request cancelled".into())),
        _ = lease.inner.stopping.cancelled() => Err(shutting_down()),
        _ = tokio::time::sleep(policy.timeout) => Err(AppError::Internal("Git process timed out; verify repository/remote state before retrying".into())),
    };
    if result.is_err() {
        let _ = child.start_kill();
        match tokio::time::timeout(Duration::from_secs(2), child.wait()).await {
            Ok(Ok(_)) => lease.reaped.store(true, Ordering::SeqCst),
            _ => {
                return Err(AppError::Internal(
                    "Git child cleanup was not confirmed; outcome requires verification".into(),
                ));
            }
        }
    } else {
        lease.reaped.store(true, Ordering::SeqCst);
    }
    result
}

File: apps/desktop/src/core/git/command.ts
import { type ChildProcess, spawn } from "node:child_process";
import { restrictedEnvironment } from "./access";
import { conflict, internalError, sanitize } from "./support";

/**
 * The one place a Git child process is started.
 *
 * A port of the pre-merge implementation, which spawns Git the same way
 * for different owners. What both establish, and what this keeps:
 *
 *   * **The environment is a lockdown, not a suggestion.** Every variable that
 *     could redirect Git at another repository, another config or an external
 *     helper is removed, and the prompt / askpass / pager settings are fixed.
 *     A caller may add variables (the interactive rebase's sequence editor is
 *     the only one that does) and they are applied *last*, so nothing a caller
 *     passes can weaken the list above it.
 *   * **Output is bounded.** stdout and stderr each have a ceiling; passing it
 *     is an error rather than a truncated answer that reads as success.
 *   * **A cancelled or timed-out child is killed and then waited for.** The
 *     outcome of such a command is unknown, never "did not happen", and the
 *     caller is told exactly that.
 */

const STDOUT_LIMIT = 8 * 1024 * 1024;
const STDERR_LIMIT = 64 * 1024;
/** The repository service's own, smaller ceilings. */
export const REPOSITORY_STDOUT_LIMIT = 4 * 1024 * 1024;
export const REPOSITORY_STDERR_LIMIT = 64 * 1024;

/** How many Git children this process will own at once. */
const MAX_CONCURRENT = 64;

/** Variables removed outright before Git runs. */
const OVERRIDES = [
  "GIT_DIR",
  "GIT_WORK_TREE",
  "GIT_COMMON_DIR",
  "GIT_INDEX_FILE",
  "GIT_OBJECT_DIRECTORY",
  "GIT_ALTERNATE_OBJECT_DIRECTORIES",
  "GIT_NAMESPACE",
  "GIT_CONFIG",
  "GIT_CONFIG_COUNT",
  "GIT_CONFIG_PARAMETERS",
  "GIT_EXTERNAL_DIFF",
  "GIT_DIFF_OPTS",
  "GIT_CURL_VERBOSE",
];

const PREFIXES = ["GIT_TRACE", "GIT_CONFIG_KEY_", "GIT_CONFIG_VALUE_"];

export interface GitEnvironmentOptions {
  /** Applied after the lockdown; the rebase sequence editor only. */
  readonly extra?: Readonly<Record<string, string>>;
  /** Without the execution grant, helpers and lazy fetches are refused too. */
  readonly restrict?: boolean;
  /**
   * The askpass helper `core/remote` publishes, so an `ssh://` remote can ask
   * a person for a passphrase instead of failing on a missing TTY. Absent means
   * no helper is configured and both askpass variables stay empty — which is
   * what refuses the prompt rather than hanging on it.
   */
  readonly askpass?: string;
}

/**
 * The environment one Git child runs with.
 *
 * Built from `ambient` rather than from scratch: Git needs `PATH`, `HOME` and
 * the platform's own variables to find its exec path and the user's config.
 * What it must not inherit is anything in {@link OVERRIDES} or matching
 * {@link PREFIXES}.
 */
export function gitEnvironment(
  options: GitEnvironmentOptions = {},
  ambient: NodeJS.ProcessEnv = process.env,
): NodeJS.ProcessEnv {
  const environment: NodeJS.ProcessEnv = { ...ambient };
  for (const key of Object.keys(environment)) {
    if (
      OVERRIDES.includes(key) ||
      PREFIXES.some((prefix) => key.startsWith(prefix))
    ) {
      delete environment[key];
    }
  }
  environment.GIT_TERMINAL_PROMPT = "0";
  environment.GIT_ASKPASS = options.askpass ?? "";
  environment.SSH_ASKPASS = options.askpass ?? "";
  if (options.askpass !== undefined) environment.SSH_ASKPASS_REQUIRE = "force";
  environment.GCM_INTERACTIVE = "never";
  environment.GIT_OPTIONAL_LOCKS = "0";
  environment.LC_ALL = "C";
  if (options.restrict === true) {
    for (const [key, value] of Object.entries(restrictedEnvironment())) {
      if (value === undefined) delete environment[key];
      else environment[key] = value;
    }
  }
  for (const [key, value] of Object.entries(options.extra ?? {})) {
    environment[key] = value;
  }
  return environment;
}

/** The fixed prefix every legacy-surface Git command carries. */
export const LEGACY_PREFIX = [
  "--no-pager",
  "-c",
  "core.quotepath=false",
  "-c",
  "color.ui=false",
];

/** The repository service adds one more: a fixed log output encoding. */
export const REPOSITORY_PREFIX = [
  "--no-pager",
  "-c",
  "core.quotepath=false",
  "-c",
  "color.ui=false",
  "-c",
  "i18n.logOutputEncoding=UTF-8",
];

export interface CommandOutput {
  readonly stdout: Buffer;
  readonly stderr: Buffer;
  /** `null` when the child was killed by a signal rather than exiting. */
  readonly status: number | null;
}

export interface RunOptions {
  readonly cwd: string;
  readonly args: readonly string[];
  readonly prefix?: readonly string[];
  readonly timeoutMs: number;
  readonly environment?: GitEnvironmentOptions;
  readonly stdoutLimit?: number;
  readonly stderrLimit?: number;
  /** Bytes written to the child's stdin; absent closes it immediately. */
  readonly input?: Buffer;
  /** Called with each completed stderr line, for `--progress` counters. */
  readonly onStderrLine?: (line: string) => void;
  /** Set as soon as the child exists: from here the outcome is unknown. */
  readonly onStarted?: () => void;
  readonly signal?: AbortSignal;
}

let active = 0;
let stopping = false;
let cleanupFailed = false;
const running = new Set<ChildProcess>();

/** How many Git children this process currently owns. */
export function activeCommands(): number {
  return active;
}

/**
 * Refuse new commands, kill the ones in flight and wait for them to be reaped.
 *
 * `shutdown_legacy_operations` in Rust; the same contract, including that an
 * unconfirmed cleanup is reported rather than swallowed.
 */
export async function shutdown(timeoutMs: number): Promise<void> {
  stopping = true;
  for (const child of running) child.kill("SIGKILL");
  const deadline = Date.now() + timeoutMs;
  while (running.size > 0) {
    if (Date.now() > deadline) {
      throw internalError(
        "Some Git processes did not stop before the deadline",
      );
    }
    await sleep(20);
  }
  if (cleanupFailed) {
    throw internalError("Git child cleanup requires inspection");
  }
}

/** Lets a test start from a clean slate after exercising {@link shutdown}. */
export function resume(): void {
  stopping = false;
  cleanupFailed = false;
}

export async function runGit(options: RunOptions): Promise<CommandOutput> {
  if (stopping) throw conflict("Git service is shutting down");
  if (active >= MAX_CONCURRENT) {
    throw conflict("Too many active Git processes");
  }
  if (options.signal?.aborted === true) {
    throw conflict("Git process cancelled before launch");
  }
  active += 1;
  try {
    return await capture(options);
  } finally {
    active -= 1;
  }
}

function capture(options: RunOptions): Promise<CommandOutput> {
  const stdoutLimit = options.stdoutLimit ?? STDOUT_LIMIT;
  const stderrLimit = options.stderrLimit ?? STDERR_LIMIT;
  const child = spawn(
    "git",
    [...(options.prefix ?? LEGACY_PREFIX), ...options.args],
    {
      cwd: options.cwd,
      env: gitEnvironment(options.environment),
      stdio: [options.input === undefined ? "ignore" : "pipe", "pipe", "pipe"],
      windowsHide: true,
    },
  );
  running.add(child);
  options.onStarted?.();

  return new Promise<CommandOutput>((resolve, reject) => {
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    let stdoutBytes = 0;
    let stderrBytes = 0;
    let pending = "";
    let settled = false;
    let exit: { code: number | null } | undefined;
    let closedStreams = 0;

    const finish = (error?: Error): void => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      options.signal?.removeEventListener("abort", onAbort);
      if (error === undefined) {
        running.delete(child);
        resolve({
          stdout: Buffer.concat(stdout),
          stderr: Buffer.concat(stderr),
          status: exit?.code ?? null,
        });
        return;
      }
      // The outcome of a killed command is unknown, so the kill is confirmed
      // before the caller is told anything at all.
      child.kill("SIGKILL");
      const reap = setTimeout(() => {
        cleanupFailed = true;
        running.delete(child);
        reject(internalError("Git process cleanup was not confirmed"));
      }, 2_000);
      child.once("close", () => {
        clearTimeout(reap);
        running.delete(child);
        reject(error);
      });
    };

    const timer = setTimeout(() => {
      finish(
        internalError(
          "Git process timed out; its outcome requires verification before retrying",
        ),
      );
    }, options.timeoutMs);
    const onAbort = (): void => {
      finish(
        conflict(
          "Git process cancelled; its outcome requires verification before retrying",
        ),
      );
    };
    options.signal?.addEventListener("abort", onAbort, { once: true });

    child.stdout?.on("data", (chunk: Buffer) => {
      stdoutBytes += chunk.byteLength;
      if (stdoutBytes > stdoutLimit) {
        finish(
          internalError(
            "Git output exceeded its bounded budget; verify repository state before retrying",
          ),
        );
        return;
      }
      stdout.push(chunk);
    });
    child.stderr?.on("data", (chunk: Buffer) => {
      if (options.onStderrLine !== undefined) {
        // Git writes progress with carriage returns rather than newlines, so
        // both count as a line end. The buffer is bounded the same way the
        // read is: a remote that never writes a separator cannot grow it.
        pending += chunk.toString("utf8");
        let index = pending.search(/[\r\n]/);
        while (index >= 0) {
          options.onStderrLine(pending.slice(0, index));
          pending = pending.slice(index + 1);
          index = pending.search(/[\r\n]/);
        }
        if (pending.length > 4096) pending = pending.slice(-4096);
      }
      stderrBytes += chunk.byteLength;
      if (stderrBytes > stderrLimit) {
        // stderr is progress as often as it is an error; the progress reader
        // keeps only a bounded tail rather than failing the command.
        if (options.onStderrLine !== undefined) {
          stderr.push(chunk);
          let total = stderr.reduce((sum, part) => sum + part.byteLength, 0);
          while (total > stderrLimit && stderr.length > 1) {
            total -= (stderr.shift() as Buffer).byteLength;
          }
          return;
        }
        finish(
          internalError(
            "Git output exceeded its bounded budget; verify repository state before retrying",
          ),
        );
        return;
      }
      stderr.push(chunk);
    });

    const streamClosed = (): void => {
      closedStreams += 1;
      if (closedStreams === 2 && exit !== undefined) {
        if (pending !== "") options.onStderrLine?.(pending);
        finish();
      }
    };
    child.stdout?.once("end", streamClosed);
    child.stderr?.once("end", streamClosed);

    child.once("error", () => {
      settled = true;
      clearTimeout(timer);
      running.delete(child);
      reject(internalError("Could not start Git"));
    });
    child.once("exit", (code) => {
      exit = { code };
      if (closedStreams === 2) {
        if (pending !== "") options.onStderrLine?.(pending);
        finish();
      }
    });

    if (options.input !== undefined && child.stdin !== null) {
      child.stdin.on("error", () => {});
      child.stdin.end(options.input);
    }
  });
}

/** `Receiving objects:  47% (470/1000)` → 47, or `undefined`. */
export function progressPercent(line: string): number | undefined {
  const marker = line.indexOf("%");
  if (marker < 0) return undefined;
  let start = marker;
  while (start > 0 && /[0-9]/.test(line[start - 1] as string)) start -= 1;
  if (start === marker) return undefined;
  const percent = Number.parseInt(line.slice(start, marker), 10);
  return Number.isNaN(percent) || percent > 100 ? undefined : percent;
}

/** The refusal a non-zero exit becomes on the legacy surface. */
export function commandFailure(output: CommandOutput): Error {
  return internalError(sanitize(output.stderr.toString("utf8")));
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
